### PR TITLE
Save cluster snapshots and recoveries in the local storage

### PR DESCRIPTION
**Pull request by zhussupov-t**
Save snapshots of cluster discovery and recoveries data. Using sqlite as storage.

### CODE REVIEW

File: internal/config/config.go [hunk 1]
```diff
@@ -54,6 +54,11 @@ type Config struct {
 			Timeout                  time.Duration `yaml:"timeout"`
 			TimeoutAsync             time.Duration `yaml:"timeout_async"`
 		} `yaml:"hooks"`
+		Storage struct {
```
Comment (pparshin) [inline internal/config/config.go:57]: You should provide default values and add some unit tests

File: config/qumomf.conf.yml [hunk 2]
```diff
@@ -66,6 +66,12 @@ qumomf:
     post_unsuccessful_failover:
       - "echo 'Failed to recover from {failureType} on {failureCluster}. Set: {failureReplicaSetUUID}; Failed: {failedURI}' >> /tmp/qumomf_recovery.log"
 
+  # Sqllite storage to save snapshots and recoveries information
```
Comment (pparshin) [inline config/qumomf.conf.yml:69]: Local persistent storage to save snapshots, recoveries and other useful data

File: cmd/qumomf/main.go [hunk 3]
```diff
@@ -59,7 +60,12 @@ func main() {
 		logger.Warn().Msg("No clusters are found in the configuration")
 	}
 
-	qCoordinator := coordinator.New(logger)
+	relStorage, err := newStorage(cfg)
```
Comment (pparshin) [inline cmd/qumomf/main.go:63]: I would rename to simple `db`

File: internal/storage/storage.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	ErrEmptyResult = errors.New("empty result")
+)
+
+type SaveRequest struct {
+	ClusterName string
+	CreatedAt   int64
+	Data        []byte
+}
+
+type Storage interface {
```
Comment (pparshin) [inline internal/storage/storage.go:50]: You have an interface of your repository. Move the concrete implementation to another file `sqlite.go` and add some constructors

```go
type Storage interface { ... }

func NewSQLiteStorage(cfg Config) (*SQLiteStorage, error) { ... }

func NewMockStorage() (*MockStorage, error) { ... }
```

You might move implementations to its own packages.

File: internal/storage/storage.go [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	SaveSnapshot(context.Context, SaveRequest) error
+	SaveRecovery(context.Context, SaveRequest) error
+	GetClusterLastSnapshot(context.Context, string) ([]byte, error)
```
Comment (pparshin) [inline internal/storage/storage.go:53]: It is quite strange to return raw bytes instead of DTO

File: internal/storage/storage.go [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+	ConnectTimeout time.Duration
+	QueryTimeout   time.Duration
+}
+
+func NewStorage(cfg Config) (Storage, error) {
+	cfg = validateConfig(cfg)
+	ctx, cancel := context.WithTimeout(context.Background(), cfg.ConnectTimeout)
+	defer cancel()
+
+	_, err := createFileIfNotExists(cfg.FileName)
```
Comment (pparshin) [inline internal/storage/storage.go:73]: I suppose `sql.Open` already checks and creates the file if it is not exist

File: internal/storage/storage.go [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+	if err != nil {
+		return &storage{}, nil
+	}
+
+	db, err := sql.Open("sqlite3", cfg.FileName)
+	if err != nil {
+		return &storage{}, err
```
Comment (pparshin) [inline internal/storage/storage.go:80]: Return `nil` instead of storage instance

File: internal/storage/storage.go [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	return cfg
+}
+
+func (s *storage) SaveSnapshot(ctx context.Context, sr SaveRequest) error {
+	ctx, cancel := context.WithTimeout(ctx, s.config.QueryTimeout)
+	defer cancel()
+
+	stmt, err := s.db.PrepareContext(ctx, querySaveSnapshot)
```
Comment (pparshin) [inline internal/storage/storage.go:116]: Do you need  the prepared statement? Just use `s.db.ExecContext`

File: internal/storage/storage.go [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+		}
+
+		resp = append(resp, data)
+	}
+
+	return resp, err
+}
+
+func createTables(ctx context.Context, db *sql.DB) error {
+	for _, q := range []string{queryCreateTableRecoveries, queryCreateTableSnapshots} {
```
Comment (pparshin) [inline internal/storage/storage.go:189]: Move all database initialization queries to the single one:

```go
initDatabaseQueries = `
  ...
`
```

File: internal/vshard/cluster.go [hunk 10]
```diff
@@ -99,9 +99,11 @@ type Cluster struct {
 
 	mutex  sync.RWMutex
 	logger zerolog.Logger
+
+	onClusterDiscoveredCB func(string, int64, []byte)
 }
 
-func NewCluster(name string, cfg config.ClusterConfig) *Cluster {
+func NewCluster(name string, onClusterDiscovered func(string, int64, []byte), cfg config.ClusterConfig) *Cluster {
```
Comment (pparshin) [inline internal/vshard/cluster.go:106]: Callbacks should be setting by setters. Do not extend the constructor - it is an optional feature.

File: internal/vshard/cluster.go [hunk 11]
```diff
@@ -373,10 +376,23 @@ func (c *Cluster) Discover() {
...
+
+		go c.saveSnapshotToStorage(ns.Copy())
 	}
 	c.mutex.Unlock()
 }
 
+func (c *Cluster) saveSnapshotToStorage(ns Snapshot) {
+	data, err := json.Marshal(ns)
+	if err != nil {
+		c.logger.Error().Err(err).Msg("failed to marshall snapshot data")
```
Comment (pparshin) [inline internal/vshard/cluster.go:388]: `Error().Err(err)` can be simplified to `.Err(err)`

File: internal/vshard/cluster.go [hunk 12]
```diff
@@ -373,10 +376,23 @@ func (c *Cluster) Discover() {
 	if c.snapshot.Created <= ns.Created {
 		ns.UpdatePriorities(c.snapshot.priorities)
 		c.snapshot = ns
+
+		go c.saveSnapshotToStorage(ns.Copy())
 	}
 	c.mutex.Unlock()
 }
 
+func (c *Cluster) saveSnapshotToStorage(ns Snapshot) {
```
Comment (pparshin) [inline internal/vshard/cluster.go:385]: You do not know anything about what callback does, so just call in appropriate place:

```go
if c.onClusterDiscovered != nil {
  c.onClusterDiscovered()
}
```

**Review by pparshin — CHANGES_REQUESTED** [review]

File: internal/storage/storage.go (hunk 5 above)
Comment (zhussupov-t) [reply internal/storage/storage.go]: Replaced with DTO

**Comment by pparshin** [comment]
Closes #24

File: cmd/qumomf/main.go [hunk 13]
```diff
@@ -59,7 +62,12 @@ func main() {
 		logger.Warn().Msg("No clusters are found in the configuration")
 	}
 
-	qCoordinator := coordinator.New(logger)
+	db, err := newStorage(cfg)
+	if err != nil {
+		logger.Fatal().Err(err).Msg("failed to init sqlite storage")
```
Comment (pparshin) [inline cmd/qumomf/main.go:67]: Do not clarify the concrete storage, just 'failed to init persistent storage'. Any sqlite mentions should be in the error message

File: cmd/qumomf/main.go [hunk 14]
```diff
@@ -82,6 +90,14 @@ func main() {
 	}
 }
 
+func newStorage(cfg *config.Config) (storage.Storage, error) {
+	return sqlite.NewSQLiteStorage(sqlite.Config{
```
Comment (pparshin) [inline cmd/qumomf/main.go:94]: `sqlite.New`

File: config/qumomf.conf.yml [hunk 15]
```diff
@@ -66,6 +66,12 @@ qumomf:
     post_unsuccessful_failover:
       - "echo 'Failed to recover from {failureType} on {failureCluster}. Set: {failureReplicaSetUUID}; Failed: {failedURI}' >> /tmp/qumomf_recovery.log"
 
+  # Local persistent storage to save snapshots, recoveries and other useful data
+  storage:
+    filename: 'sqlite.db'
```
Comment (pparshin) [inline config/qumomf.conf.yml:71]: `qumomf.db`

File: internal/config/config.go [hunk 16]
```diff
@@ -31,6 +31,9 @@ const (
 	defaultAsyncHookTimeout          = 10 * time.Minute
 	defaultMaxFollowerLSNLag         = 1000
 	defaultMaxFollowerIdle           = 5 * time.Minute
+	defaultStorageFileName           = "sqlite.db"
```
Comment (pparshin) [inline internal/config/config.go:34]: `qumomf.db`

File: internal/coordinator/coordinator.go [hunk 17]
```diff
@@ -45,6 +50,7 @@ func (c *Coordinator) RegisterCluster(name string, cfg config.ClusterConfig, glo
 
 	cluster := vshard.NewCluster(name, cfg)
 	cluster.SetLogger(clusterLogger)
+	cluster.SetOnClusterDiscoveredCB(c.onClusterDiscovered)
```
Comment (pparshin) [inline internal/coordinator/coordinator.go:53]: `SetOnClusterDiscoveredCB` -> `SetOnClusterDiscovered`

File: internal/storage/sqlite/sqlite.go [hunk 18]
```diff
@@ -0,0 +1,159 @@
...
+
+	"github.com/shmel1k/qumomf/internal/vshard"
+
+	"github.com/shmel1k/qumomf/internal/storage"
+)
+
+const (
+	querySaveSnapshot = `INSERT INTO snapshots(cluster_name, created_at, data) 
+							VALUES(?, ?, ?)`
+	querySaveRecoveries = `INSERT INTO recoveries(cluster_name, created_at, data)
```
Comment (pparshin) [inline internal/storage/sqlite/sqlite.go:20]: I am afraid that it is not a good idea to save all cluster snapshots, especially without expiration. It is enough to store the last one, isn't it?

File: internal/storage/storage.go [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+
+import (
+	"context"
+
+	"github.com/shmel1k/qumomf/internal/vshard/orchestrator"
+
+	"github.com/shmel1k/qumomf/internal/vshard"
+
+	// sqlite3 driver
+	_ "github.com/mattn/go-sqlite3"
```
Comment (pparshin) [inline internal/storage/storage.go:11]: Why is it imported here, not in `sqlite` package?

**Review by pparshin — CHANGES_REQUESTED** [review]

File: internal/storage/sqlite/sqlite.go (hunk 18 above)
Comment (zhussupov-t) [reply internal/storage/sqlite/sqlite.go:24]: changed the logic to update the current snapshot

**Review by pparshin — APPROVED** [review]

**Pull request merged**